src/main/resources/avro/sequenceAnnotations.avdl

```diff
@@ -0,0 +1,137 @@
+@namespace("org.ga4gh.models")
+/**
+This protocol defines annotations on GA4GH genomic sequences It includes two
+types of annotations: continuous and discrete hierarchical.
+
+The discrete hierarchical annotations are derived from the Sequence Ontology
+(SO) and GFF3 work 
+
+   http://www.sequenceontology.org/gff3.shtml
+
+The goal is to be able to store annotations using the GFF3 and SO conceptual
+model, although there is not necessarly a one-to-one mapping in Avro records
+to GFF3 records.
+
+The minimum requirement is to be able to accurately represent the current
+state of the art annotation data and the full SO model.  Feature is the
+core generic record which corresponds to the a GFF3 record.
+*/
+protocol SequenceAnnotations {
+
+  import idl "common.avdl";
+  import idl "metadata.avdl";
+
+  /**
+  Type defining a collection of attributes associated with various protocol
+  records.  Each attribute is a name that maps to an array of one or more
+  values.  Values can be strings, external identifiers, or ontology terms.
+  Values should be split into the array elements instead of using a separator
+  syntax that needs to parsed.
+  */
+  // TODO: how are multiple instances of a given attribute vs multiple values
+  // for an attribute distinguished
+  record Attributes {
```

Comment:
The definition seems clear to me from a data structure
prospective (map of lists of polymorphic values) but I am
concerned it will be convoluted to use in practice.  This was
patterned on the `map<array<string>> info' approach, only
extending it to multiple types.

VariantSetMetadata, which seems to fill the a similar role, has a
very different structure to either of these approaches.  It's
uses is not clearly documented, I really don't understand it.

I completely agree that this is not limited to
SequenceAnnotations. We needed to have a single way to specify
the extended attributes in a consistent way across the API.

Given the constraints of not having a polymorphic system in avro
or protobuf, I am not sure what is the best way to encode data
types that are fundamental GA4GH types (like ExternalIdentifier
and OntologyTerm), and for that matter int and float, in a
arbitrary key/value storage.  Multiple parallel arrays seems
wrong, although maybe pragmatic.

I will start a ticket on this issue.  However, it would be great
to move ahead on the sequence annotations alpha, as we will
learn my using it.  Any objections?


CH Albach <notifications@github.com> writes:
> My initial impression is that this data type is convoluted; by default, I would
> assume this type of field should follow the same pattern we have everywhere
> else in the API i.e. map<[]string>.
>
> In particular, referencing an
> ExternalIdentifier does not seem like a problem unique to sequence annotations.
> I'd guess this is probably meant to emulate the behavior of GFF3, but given the
> current documentation I can't tell if this is capturing some semantic nuance in
> the spec, or if it's overfitting and this is an artifact of it being a file
> format.
> 
> This could continue to follow the normal API patterns we have elsewhere by
> splitting this into three fields on the FeatureSet:
> 
> map<array<string>> info;
> array<ExternalIdentifier> externalIds;
> map<array<OntologyTerm>> ___;
> 
> I don't understand the last use case well enough to propose a name for that
> field.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub*
>